core/dbt/graph/selector.py

```diff
@@ -162,12 +187,53 @@ def filter_selection(self, selected: Set[UniqueId]) -> Set[UniqueId]:
             unique_id for unique_id in selected if self._is_match(unique_id)
         }
 
-    def expand_selection(self, selected: Set[UniqueId]) -> Set[UniqueId]:
-        """Perform selector-specific expansion."""
+    def expand_selection(
+        self, selected: Set[UniqueId], greedy: bool = False
+    ) -> Tuple[Set[UniqueId], Set[UniqueId]]:
+        # Test selection can expand to include an implicitly/indirectly selected test.
+        # In this way, `dbt test -m model_a` also includes tests that directly depend on `model_a`.
+        # Expansion has two modes, GREEDY and NOT GREEDY.
+        #
+        # GREEDY mode: If ANY parent is selected, select the test. We use this for EXCLUSION.
+        #
+        # NOT GREEDY mode:
+        #  - If ALL parents are selected, select the test.
+        #  - If ANY parent is missing, return it separately. We'll keep it around
+        #    for later and see if its other parents show up.
+        # We use this for INCLUSION.
+
+        direct_nodes = selected
+        indirect_nodes = set()
+
+        for unique_id in self.graph.select_successors(selected):
+            if unique_id in self.manifest.nodes:
+                node = self.manifest.nodes[unique_id]
+                if node.resource_type == NodeType.Test:
```

Comment:
if there's a single thing that i don't feel great about in this PR, it's this line! I think that the majority of node selection is resource type-agnostic, and the incidence of Test in here is setting off all of my "weird abstraction" alarm bells. I don't know if the thing to do is:
1. Make tests less weird / not a node type, such that this logic wouldn't need to exist (but what?)
2. Make selection logic resource type-specific, such that selectors do fundamentally different things for different node types (this feels bad)

So, not claiming this as a blocker or challenge to overcome - I think this change is behavior is a real net-positive, but it would be good for us to come up with an abstraction over / reimagination of tests that doesn't _require_ this kind of logic to exist!

Reply:
Yeah, this is a really fair point. The goal of this PR was to make selection _more_ consistent overall, while at the same time acknowledging that (a) there exists this special thing, which I'm calling "indirect selection," and (b) only test nodes qualify for it.

I formalized that by adding a super-basic function to the `selector` module, `can_select_indirectly`, which returns `True` for tests and `False` otherwise. For now, it's just tests, but if we ever added another resource type that's sort-of-a-node, sort-of-an-attribute-of-another-node, we could include it in this function. And if we decided to introduce some kind of runtime variable / flag that turned on indirect selection for everything, that'd be the place to do it.